### ISSUE TITLE
AutoValue: Support Optional<T> similar to @Nullable

### ISSUE THREAD

**Issue by mfulton26** [google/auto#278]
I'd like to have "optional" value properties but instead of annotating my properties with `@Nullable` and then checking for `null` when using the property I'd rather use [`java.util.Optional<T>`](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html) and/or [`com.google.common.base.Optional<T>`](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/base/Optional.html).

Here's what I'm doing today:
```java
import com.google.auto.value.AutoValue;

import java.util.Optional;

class Example {
  @AutoValue
  abstract static class Animal {
    static Builder builder() {
      return new AutoValue_Example_Animal.Builder()
          .name(Optional.empty)
          .numberOfLegs(Optional.empty());
    }

    abstract Optional<String> name();
    abstract Optional<Integer> numberOfLegs();

    @AutoValue.Builder
    abstract static class Builder {
      Builder name(String s) { return name(Optional.ofNullable(s)); }
      abstract Builder name(Optional<String> s);
      Builder name(Integer n) { return name(Optional.ofNullable(n)); }
      abstract Builder numberOfLegs(Optional<Integer> n);
      abstract Animal build();
    }
  }
}
```

Here's what I want to do:
```java
import com.google.auto.value.AutoValue;

import java.util.Optional;

class Example {
  @AutoValue
  abstract static class Animal {
    static Builder builder() {
      return new AutoValue_Example_Animal.Builder();
    }

    abstract Optional<String> name();
    abstract Optional<Integer> numberOfLegs();

    @AutoValue.Builder
    abstract static class Builder {
      abstract Builder name(String s);
      abstract Builder numberOfLegs(Integer n);
      abstract Animal build();
    }
  }
}
```

i.e. I want `Optional<T>` properties to default to [`Optional.empty()`](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#empty--)/[`Optional.absent()`](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/base/Optional.html#absent()) instead of `null` and to be able to define `Builder` methods (or constructors) that use `T` instead of `Optional<T>`. I'm using [`Optional.ofNullable(T)`](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#ofNullable-T-)/[`Optional.fromNullable(T)`](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/base/Optional.html#fromNullable(T)) instead of [`Optional.of(T)`](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#of-T-)/[`Optional.of(T)`](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/base/Optional.html#of(T)) so that I can still set a property to [`Optional.empty()`](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#empty--)/[`Optional.absent()`](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/base/Optional.html#absent()) by passing in `null` to the appropriate `Builder` method.

**Comment by eamonnmcmanus** [comment]
Yes, we already have a Google-internal issue tracking this.

**Comment by eamonnmcmanus** [comment]
I should say that opinion seems to be against the idea of allowing null to mean setting the Optional to empty.

**Comment by mfulton26** [comment]
True, `Optional.empty() != null` and shouldn't. I was just hoping to avoid needing to specify multiple methods on the `Builder` for a single property... `abstract Builder nameFromNullable(String)`) seems odd to me (presently).

Out of curiosity (as I can't see the Google-internal issue tracking this) what are some of the other options? Define `abstract Builder name(String)` and `abstract Builder name(Optional<String>)` if you want both?

**Comment by eamonnmcmanus** [comment]
I think specifying both `name(String)` and `name(Optional<String>)` would make good sense if you want to be able to say `name(Optional.empty())`, yes.

**Comment by mfulton26** [comment]
:thumbsup:

**Comment by hgschmie** [comment]
Any progress planned here? I am trying to adopt auto value and not being able to deal with Optional<T> properties without glue code as described by @mfulton26 is definitely hindering it.

**Comment by hgschmie** [comment]
Let me add my second comment here as you chose to simply close #300 without actually thinking it through:

Yes, as you point out in your code example, I want to have access to the "raw" state of the property in the builder without having the getter penalizing me with an IllegalStateException. Having an `Optional<T>` getter for a property of `T` is fine, but it will get a bit awkward because Beans with "normal" attributes and "optional" attributes would (should) look the same in the Builders:

```java
@AutoValue
class Bean1 {
  String getFoo();

  public static Builder builder() { return new AutoValue_Bean1.Builder(); }

  @AutoValue.Builder
  abstract static class Builder {
    abstract Builder foo(String foo);

    abstract Optional<String> getFoo();

    Bean1 build() {
      if (getFoo().isPresent()) {
       ...
      }
     return autoBuild();
  }
}
```

and 

```java
@AutoValue
class Bean1 {
  Optional<String> getFoo();

  public static Builder builder() { return new AutoValue_Bean1.Builder(); }

  @AutoValue.Builder
  abstract static class Builder {
    abstract Builder foo(Optional<String> foo);

    abstract Optional<String> getFoo();

    Bean1 build() {
      if (getFoo().isPresent()) {
       ...
      }
     return autoBuild();
  }
}
```

This use would require to preset `Optional` properties with `Optional.empty()` / `Optional.absent()` values. Otherwise, you end up with this:

```java
@AutoValue
class Bean1 {
  Optional<String> getFoo();

  public static Builder builder() { return new AutoValue_Bean1.Builder(); }

  @AutoValue.Builder
  abstract static class Builder {
    abstract Builder foo(Optional<String> foo);

    abstract Optional<Optional<String>> getFoo();

    Bean1 build() {
      if (getFoo().isPresent()) {
       ...
      }
     return autoBuild();
  }
}
```

I assume that no one wants that. :-)

**Comment by eamonnmcmanus** [comment]
In the second extract from the preceding comment, if you have a property of type `Optional<String>` we will give it a default value of `Optional.empty()`. The special `Optional<T>` wrapping to determine if properties have been set only makes sense for properties that don't have a default value, namely primitive properties or reference properties that are neither `@Nullable` nor `Optional`. Since in the example `getFoo()` is indeed `Optional`, there will be no way to determine whether it was explicitly set to `Optional.empty()` or left as that default value. In my opinion, making such a distinction would be just as confusing as the `Optional<Optional<String>>` type implies.

**Comment by eamonnmcmanus** [comment]
This is implemented in the latest snapshot.

**Comment by kenzierocks** [comment]
I'm really glad this support has been added, but do you think it would be possible to make generated methods whose `super` has `@Nullable` on the parameter use `ofNullable` rather than `of`? This way I could "default" to a certain value but other code could set the value to be empty, or in the case of a `with` method, to change values already set.

**Comment by eamonnmcmanus** [comment]
Sorry, I'm not really clear on what you're asking for. Could you given an example?

**Comment by kenzierocks** [comment]
Something like [this](https://gist.github.com/kenzierocks/53fcc433984db0fafae9be65af6c29f9).

If I do `Value.builder().foo(null)`, I want `builder.build().getFoo()` to return `Optional.empty()`.

**Comment by eamonnmcmanus** [comment]
OK, that's clearer now. To summarize your example here:

```java
@AutoValue
abstract class Value {
  public abstract Optional<String> foo();

  @AutoValue.Builder
    interface Builder {
        Builder setFoo(@Nullable String value);
        Value build();
    }
}
```

You want `builder.setFoo(null)` to set `foo` to `Optional.empty()`. That seems like a reasonable request.

**Comment by tetv** [comment]
Typically the best practices says that Option<T> shouldn't be used as fields or parameters.
One of the reasons is that Optional is not serializable by design.
Therefore my suggestion is that when there is:
`public abstract Optional<String> foo();`
..should be translated to:
`private String foo;
public Optional<String> foo() { return Optional.ofNullable(foo); }`

**Comment by jbgi** [comment]
@tetv So, only if the annotated class extends ```Serializable``` then?

**Comment by tetv** [comment]
My suggestion is that all the generated classes (value or builder) shouldn't have fields or parameters of type `Optional<T>`. But yes, if the class extends `Serializable` it's a must not having them at all.

Note that Optional was designed to provide a limited mechanism for library method return types where there needed to be a clear way to represent "no result".

**Comment by eamonnmcmanus** [comment]
I think most people would expect that if you have an `@AutoValue` method `abstract Foo bar()` then the generated class will have a property of type `Foo`. If `Foo` is actually `Optional<Baz>` and if your class is `Serializable` then your class will not actually be serializable. That is the same as if `Foo` were any other unserializable type. We could indeed work around this particular case by using null to mean absent, but I'm not sure we would want to. Apart from anything else, it would mean there would be a hidden cost, since we would need to recreate the `Optional<Baz>` every time `bar()` was called, or introduce a second volatile field to store it.

**Comment by tetv** [comment]
If `Foo` doesn't extend `Serializable` and the `@AutoValue` class does it, then the `Foo` field should be transient. Note that it make all sense a value class be `Serializable` if specified.

**Comment by tetv** [comment]
Some thread about:

1) Why `Optional<T>` shouldn't be used as fields or method parameters:
[http://blog.codefx.org/java/stephen-colebourne-optional-a-strict-approach/](http://blog.codefx.org/java/stephen-colebourne-optional-a-strict-approach/)

2) Why `Optional<T>` isn't `Serializable`:
[http://blog.codefx.org/java/dev/why-isnt-optional-serializable/](http://blog.codefx.org/java/dev/why-isnt-optional-serializable/)

**Comment by eamonnmcmanus** [comment]
I'm confused. It seems that link (1) in your reply contradicts what you are saying. If anything, it confirms the lack of consensus on exactly how `Optional` should be used. I don't really see a reason to change what I said [above](https://github.com/google/auto/issues/278#issuecomment-228775575).

**Comment by tetv** [comment]
Thank you @eamonnmcmanus! You are right I looked into it again and it's a bit confusing (wrong first link). But in general, there is a bit of duality of having or not `Optional` as fields and parameters.

For example Intellij's help also advice with warnings not to have then. And for me, it makes more sense what the Java Lambda team said about it, and the suggestions that the name of `Optional<T>` would be better if were `OptionalReturn<T>`. But maybe, in same cases, it makes sense to have them as fields.

In my experience, the value immutable classes are created but don't remain alive for much time, and in average, each property is consumed only once. In that case, the cost of building the `Optional` in the builder phase, in opposition to creating it when the property is consumed is not bigger.

I am using `@AutoValue` and I really need a `Serializable` entity. I also would like to have Optional<T> abstract methods, because they are very useful, and this is why I am requesting that the generated value class should always be `Serializable` compatible.

What are your thoughts about making `@AutoValue` compatible with `Serializable`?

**Comment by jbgi** [comment]
@tetv you can roll-out your own serializable Optional, or use guava's or fugue Option (https://bitbucket.org/atlassian/fugue).
You can also use a nullable field and make the abstract method for that field package private and expose only a public method that call the former, wrapping with Optional.ofNullable.

**Comment by tetv** [comment]
Thank you @jbgi. The idea of using `@AutoValue` is to code less, and make life easier, no?
I tried wrapping, but then the wrapper method needs to have another name: it's ugly! :(

What about having a way, e.g. using special Attributes, to specify the behaviour of the conversion?

````java
@AutoValue
@TrasientForNonSerializableFields
@NotUseOptionalInFields
abstract class Value extends Serializable {
  public abstract Optional<String> foo();
  public abstract OptionalInt bar();

  @AutoValue.Builder
  interface Builder {
        Builder setFoo(@Nullable String value);
        Builder setBar(int value);
        Value build();
    }
}
´´´´

**Comment by ccheptea** [comment]
Some will agree that having Optional<T> for fields is a bit messy. We're using [auto-value-node](https://github.com/ccheptea/auto-value-node). All you have to do is to add this method to your AutoValue class:

```java 
public abstract Node_<your_class_name> node();
```

then access the field like this:

```java
value.node().foo().ifPresent(fooVal -> doSomething(fooVal));
```
Here's a real code snippet from the project I work on:

```java
car.node().driver().club().contact().email()
    .ifPresent(Printer::println)
    .otherwise(() -> Printer::println("The email is missing"));
```
We've gotten rid of all null checks in our project.

Here's a more detailed explanation: http://www.ccheptea.com/2017-04-02-auto-value-node/

**Comment by bejibx** [comment]
Current ```Optional``` handling is totally unreliable on Android. We can not use ```Optional``` from Java 8 because it only available since Android 7.0 and also we kind of can't use Guava's ```Optional``` because of [dex method limit](https://www.contentful.com/blog/2014/10/30/android-and-the-dex-64k-methods-limit/).
For now I'm just writting separate method for every ```@Nullable``` field:
```java
@AutoValue
public abstract class ValueClass
{
    @NonNull
    public Optional<String> optionalString()
    {
        return Optional.fromNullable(string());
    }

    @Nullable
    abstract String string();
}
``` 
Which is kind of сontradicts one of the main ideas of this library - to write less code. Also those ```optionalSomething()``` method names is ugly. 
@tetv already mention all these problems and I'm totally agree with him - I think the only way ```Optional``` should be used - as method return values.
Unfortunately even if this behaviour will be changed in future, this will not  solve the problem with lack of Java 8 / Guava ```Optional``` on Android. Maybe there is a way to handle ```Optional```'s through extensions?

**Comment by lowasser** [comment]
I don't follow what's going on there.  Can you use Optional, or not?  You're adding your own code to use Optional, but you're saying you can't use the built in Optional handling?

**Comment by bejibx** [comment]
I can not use ```Optional``` from Java 8 and Guava. I can use my own ```Optional``` or ```Optinal``` from other libraries. Currently I'm using ```Optional``` from (Stream API)[https://github.com/aNNiMON/Lightweight-Stream-API], but AutoValue only supports ```Optional``` from Java 8 and Guava.

**Comment by kenzierocks** [comment]
It's not hard to add support for them, but it does get a little difficult because we need to know what methods to use to do the `null`<->`Optional` swap. I don't think we can just add every library's version of Optional.

**Comment by bejibx** [comment]
Thats why I asked maybe there is a way to handle ```Optional```'s with extensions? Maybe add special OptionalExtension or something?

**Comment by ccheptea** [comment]
@bejibx, I'm not sure of how badly you need to use `Optional`s, but if you could do with an alternative check out the [auto-value-node](https://github.com/ccheptea/auto-value-node) extension (I mentioned it above)? It  doesn't use any `Optional` but it resembles it and, for me at least, it solved the problem with `Nullable` fields. The integration is really simple, you just need to add a single method to your `AutoValue` class:
```java
@AutoValue
public abstract class ValueClass
{
    @Nullable
    abstract String string();
    
    abstract Node_ValueClass node();
}
```

And then you can access the field through `node()` like this:

```java
valueClassInstance.node().string()
    .ifPresent(value -> doSomeStringProcessing(value))
    .otherwise(() -> doSomethingElse());
```

The main idea is that each field is viewed as an optional node in a tree. The good part is that if you have a deep data structure you can chain the nodes to get to your field. So for the sake of an example, say `ValueClass` looks like this:

```java
@AutoValue
public abstract class ValueClass
{
    @Nullable
    abstract String string();
    
    @Nullable
    abstract AnotherClass another();
    
    abstract Node_ValueClass node();
}
```
and 

```java
@AutoValue
public abstract class AnotherClass
{
    @Nullable
    abstract String foo();

    abstract Node_ValueClass node();
}
```

now  access `foo` like this (I'm not sure a chain like this is even possible with `Optional`s):

```java
valueClassInstance.node().another().foo()
    .ifPresent(value -> doSomethingWithFoo(value))
```
if you really need to make it an `Optional` later on, use `map`:

```java
valueClassInstance.node().another().foo()
    .map(Optional::ofNullable);
```
I hope this is of some help. If not, maybe [this announcement](https://android-developers.googleblog.com/2017/03/future-of-java-8-language-feature.html) is good news.

**Comment by bejibx** [comment]
@ccheptea this extension looks great and I'll definitely try it in one of my own project. Sadly, it is hard for me to сonvince my co-workers that we need to use it in our project. It's a miracle that I was convinced them to start using AutoValue, now I'm trying to start a tradition of using ```Optional``` and, even this is a native Java language feature, I feel some serious resistance about using it. So I don't think I'll be able to use it in my work project.

> I hope this is of some help. If not, maybe this announcement is good news.

Sadly, no. ```Optional``` and other stuff (streams, functional interfaces) will only be available [since Android 7.0](https://developer.android.com/studio/preview/features/java8-support.html).

**Comment by oxygenecore** [comment]
@bejibx Optional is somewhat debatable because if should have been a language feature - with one and only way of using it. For Optional it's not that simple (Optional args? Optional fields? etc), and it's API in Java 8 is very basic.
